File: examples/dekita/.claude/hooks/reflection-reminder.py
#!/usr/bin/env python3
"""PRマージや一定アクション後に振り返りをリマインド。

Why:
    タスク完了後やセッションが長時間続いた際に振り返りを促し、
    学習機会を逃さないようにする。

What:
    - gh pr merge / git merge 成功を検出しリマインド
    - 10アクションごとに定期リマインド
    - セッション状態ファイルでアクション回数を追跡

State:
    - writes: /tmp/claude-hooks/reflection-state-{session_id}.json

Remarks:
    - 非ブロック型（リマインダー表示のみ）
    - PostToolUse:Bash フック
    - PRマージリマインドと定期リマインドは排他（マージ優先）

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#1842: get_tool_result()ヘルパー使用に統一
"""

from __future__ import annotations

import json
import re
import sys
import tempfile
from pathlib import Path

# 共通モジュール
HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))
from lib.execution import log_hook_execution
from lib.hook_input import get_tool_result
from lib.session import create_hook_context, parse_hook_input

# 振り返りリマインドの間隔（アクション回数）
REMINDER_INTERVAL_ACTIONS = 10

# セッション状態ディレクトリ
SESSION_DIR = Path(tempfile.gettempdir()) / "claude-hooks"


def get_reflection_state_file(session_id: str) -> Path:
    """Get the file path for storing reflection state.

    Args:
        session_id: The Claude session ID to scope the file.

    Returns:
        Path to session-specific reflection state file.
    """
    return SESSION_DIR / f"reflection-state-{session_id}.json"


def load_reflection_state(session_id: str) -> dict:
    """振り返り状態を読み込み

    Args:
        session_id: The Claude session ID.

    Returns:
        Reflection state dictionary.
    """
    state_file = get_reflection_state_file(session_id)
    if state_file.exists():
        try:
            return json.loads(state_file.read_text())
        except Exception:
            pass  # Best effort - corrupted state is ignored
    return {"action_count": 0, "last_reminder_action": 0, "pr_merged_count": 0}


def save_reflection_state(session_id: str, state: dict) -> None:
    """振り返り状態を保存

    Args:
        session_id: The Claude session ID.
        state: State dictionary to save.
    """
    try:
        SESSION_DIR.mkdir(parents=True, exist_ok=True)
        state_file = get_reflection_state_file(session_id)
        state_file.write_text(json.dumps(state))
    except Exception:
        # State persistence is best-effort; failures here should not
        # block the hook or affect Claude Code operation
        pass


def increment_action_count(state: dict) -> int:
    """アクションカウントをインクリメントして返す

    セッション中のBashコマンド実行回数を追跡する。
    状態ファイルに保存されたカウントを使用する。
    """
    state["action_count"] = state.get("action_count", 0) + 1
    return state["action_count"]


def is_pr_merge_command(command: str) -> bool:
    """PRマージコマンドかどうかを判定"""
    # gh pr merge パターン
    if re.search(r"gh\s+pr\s+merge", command):
        return True
    # git merge with PR branch パターン
    if re.search(r"git\s+merge.*(?:feat|fix|docs|refactor|test)/", command):
        return True
    return False


def check_pr_merge_result(tool_result: dict) -> bool:
    """PRマージが成功したかどうかを確認"""
    # 終了コードが0ならマージ成功
    if tool_result.get("exit_code", 1) != 0:
        return False

    stdout = tool_result.get("stdout", "")
    # マージ成功のメッセージを確認
    # gh pr merge: "Merged", "Pull request"
    # git merge: "Merge made by", "Fast-forward"
    merge_indicators = [
        "Merged",
        "merged",
        "Pull request",
        "Merge made by",
        "Fast-forward",
    ]
    return any(indicator in stdout for indicator in merge_indicators)


def main():
    """Remind about reflection after merge operations."""
    result = {"continue": True}

    try:
        input_data = parse_hook_input()

        ctx = create_hook_context(input_data)
        tool_name = input_data.get("tool_name", "")
        tool_input = input_data.get("tool_input", {})

        # Issue #1842: Use standardized helper for tool result extraction
        # Ensure we have a dict for .get() calls (tool_result can be a string)
        raw_result = get_tool_result(input_data)
        tool_result = raw_result if isinstance(raw_result, dict) else {}

        # Bash以外はスキップ
        if tool_name != "Bash":
            print(json.dumps(result))
            return

        command = tool_input.get("command", "")
        session_id = ctx.get_session_id()

        # 状態を読み込み（セッションIDでファイルが分離されるためリセット不要）
        state = load_reflection_state(session_id)

        reminder_message = None

        # 1. PRマージ検出
        if is_pr_merge_command(command) and check_pr_merge_result(tool_result):
            state["pr_merged_count"] = state.get("pr_merged_count", 0) + 1
            reminder_message = (
                "🎉 PRがマージされました！\n"
                "タスク完了後は振り返り（五省）を行うと効果的です:\n"
                "- 要件を正確に理解できたか\n"
                "- 実装品質は十分か\n"
                "- 検証は適切に行ったか\n"
                "- 効率的に作業できたか"
            )

        # 2. 定期リマインダー（一定回数のアクション後）
        # アクションカウントをインクリメント
        current_action_count = increment_action_count(state)
        last_reminder_count = state.get("last_reminder_action", 0)
        # REMINDER_INTERVAL_ACTIONS回ごとにリマインド
        if (
            current_action_count // REMINDER_INTERVAL_ACTIONS
            > last_reminder_count // REMINDER_INTERVAL_ACTIONS
        ):
            state["last_reminder_action"] = current_action_count
            if not reminder_message:  # PRマージメッセージがない場合のみ
                reminder_message = (
                    f"📊 セッション進行中（{current_action_count}回のアクション）\n"
                    "定期的な振り返りを推奨します。\n"
                    "ログ: .claude/logs/execution/hook-execution-*.jsonl, .claude/logs/metrics/*.jsonl"
                )

        # 状態を保存
        save_reflection_state(session_id, state)

        # リマインダーメッセージがあれば表示
        if reminder_message:
            result["systemMessage"] = reminder_message
            log_hook_execution(
                "reflection-reminder",
                "approve",
                "Reflection reminder shown",
                {"trigger": "pr_merge" if "PR" in reminder_message else "periodic"},
            )
        else:
            log_hook_execution(
                "reflection-reminder", "approve", "No reminder triggered", {"type": "no_reminder"}
            )

    except Exception:
        # フック実行の失敗でClaude Codeをブロックしない
        pass

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/reflection_reminder.ts
#!/usr/bin/env bun
/**
 * PRマージや一定アクション後に振り返りをリマインド。
 *
 * Why:
 *   タスク完了後やセッションが長時間続いた際に振り返りを促し、
 *   学習機会を逃さないようにする。
 *
 * What:
 *   - gh pr merge / git merge 成功を検出しリマインド
 *   - 10アクションごとに定期リマインド
 *   - セッション状態ファイルでアクション回数を追跡
 *
 * State:
 *   - writes: /tmp/claude-hooks/reflection-state-{session_id}.json
 *
 * Remarks:
 *   - 非ブロック型（リマインダー表示のみ）
 *   - PostToolUse:Bash フック
 *   - PRマージリマインドと定期リマインドは排他（マージ優先）
 *   - Python版: reflection_reminder.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#1842: get_tool_result()ヘルパー使用に統一
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { SESSION_DIR } from "../lib/constants";
import { getToolResult } from "../lib/input_context";
import { logHookExecution } from "../lib/logging";
import { createContext, getSessionId, parseHookInput } from "../lib/session";

const HOOK_NAME = "reflection-reminder";

// 振り返りリマインドの間隔（アクション回数）
const REMINDER_INTERVAL_ACTIONS = 10;

interface ReflectionState {
  action_count: number;
  last_reminder_action: number;
  pr_merged_count: number;
}

/**
 * Get the file path for storing reflection state.
 */
function getReflectionStateFile(sessionId: string): string {
  return `${SESSION_DIR}/reflection-state-${sessionId || "unknown"}.json`;
}

/**
 * Load reflection state from session file.
 */
function loadReflectionState(sessionId: string): ReflectionState {
  try {
    const stateFile = getReflectionStateFile(sessionId);
    if (existsSync(stateFile)) {
      const data = JSON.parse(readFileSync(stateFile, "utf-8"));
      return {
        action_count: data.action_count ?? 0,
        last_reminder_action: data.last_reminder_action ?? 0,
        pr_merged_count: data.pr_merged_count ?? 0,
      };
    }
  } catch {
    // Best effort - corrupted state is ignored
  }
  return { action_count: 0, last_reminder_action: 0, pr_merged_count: 0 };
}

/**
 * Save reflection state to session file.
 */
function saveReflectionState(sessionId: string, state: ReflectionState): void {
  try {
    mkdirSync(SESSION_DIR, { recursive: true });
    const stateFile = getReflectionStateFile(sessionId);
    writeFileSync(stateFile, JSON.stringify(state), "utf-8");
  } catch {
    // State persistence is best-effort
  }
}

/**
 * Check if command is a PR merge command.
 */
export function isPrMergeCommand(command: string): boolean {
  // gh pr merge pattern
  if (/gh\s+pr\s+merge/.test(command)) {
    return true;
  }
  // git merge with PR branch pattern
  if (/git\s+merge.*(?:feat|fix|docs|refactor|test)\//.test(command)) {
    return true;
  }
  return false;
}

/**
 * Check if PR merge was successful.
 */
export function checkPrMergeResult(toolResult: Record<string, unknown>): boolean {
  // Exit code must be 0
  const exitCode = toolResult.exit_code;
  if (exitCode !== 0) {
    return false;
  }

  const stdout = String(toolResult.stdout ?? "");
  // Check for merge success indicators
  const mergeIndicators = ["Merged", "merged", "Pull request", "Merge made by", "Fast-forward"];
  return mergeIndicators.some((indicator) => stdout.includes(indicator));
}

async function main(): Promise<void> {
  const result: Record<string, unknown> = { continue: true };

  try {
    const data = await parseHookInput();
    const ctx = createContext(data);
    const sessionId = getSessionId(ctx) ?? "unknown";

    const toolName = data.tool_name ?? "";
    const toolInput = (data.tool_input as Record<string, unknown>) ?? {};

    // Use standardized helper for tool result extraction
    const rawResult = getToolResult(data);
    const toolResult: Record<string, unknown> =
      typeof rawResult === "object" && rawResult !== null
        ? (rawResult as Record<string, unknown>)
        : {};

    // Skip non-Bash tools
    if (toolName !== "Bash") {
      console.log(JSON.stringify(result));
      return;
    }

    const command = (toolInput.command as string) ?? "";

    // Load state (file is scoped by session ID)
    const state = loadReflectionState(sessionId);

    let reminderMessage: string | null = null;

    // 1. PR merge detection
    if (isPrMergeCommand(command) && checkPrMergeResult(toolResult)) {
      state.pr_merged_count = (state.pr_merged_count ?? 0) + 1;
      reminderMessage =
        "🎉 PRがマージされました！\n" +
        "タスク完了後は振り返り（五省）を行うと効果的です:\n" +
        "- 要件を正確に理解できたか\n" +
        "- 実装品質は十分か\n" +
        "- 検証は適切に行ったか\n" +
        "- 効率的に作業できたか";
    }

    // 2. Periodic reminder (after certain number of actions)
    state.action_count = (state.action_count ?? 0) + 1;
    const currentActionCount = state.action_count;
    const lastReminderCount = state.last_reminder_action ?? 0;

    // Remind every REMINDER_INTERVAL_ACTIONS actions
    if (
      Math.floor(currentActionCount / REMINDER_INTERVAL_ACTIONS) >
      Math.floor(lastReminderCount / REMINDER_INTERVAL_ACTIONS)
    ) {
      state.last_reminder_action = currentActionCount;
      if (!reminderMessage) {
        // Only if no PR merge message
        reminderMessage = `📊 セッション進行中（${currentActionCount}回のアクション）\n定期的な振り返りを推奨します。\nログ: .claude/logs/execution/hook-execution-*.jsonl, .claude/logs/metrics/*.jsonl`;
      }
    }

    // Save state
    saveReflectionState(sessionId, state);

    // Show reminder message if any
    if (reminderMessage) {
      result.systemMessage = `[${HOOK_NAME}] ${reminderMessage}`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        "Reflection reminder shown",
        { trigger: reminderMessage.includes("PR") ? "pr_merge" : "periodic" },
        { sessionId },
      );
    } else {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        "No reminder triggered",
        { type: "no_reminder" },
        { sessionId },
      );
    }
  } catch {
    // Don't block Claude Code on hook failures
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
